File: 821-shortest-distance-to-a-character/821-shortest-distance-to-a-character.js
/**
 * @param {string} s
 * @param {character} c
 * @return {number[]}
 */
var shortestToChar = function(s, c) {
  const ans = Array(s.length)
  let pre = s.indexOf(c)
  let last = s.lastIndexOf(c)
  for (let i = 0; i <= pre; i++) {
    ans[i] = pre - i
  }
  next = pre + 1
  while (next < last) {
    while (s[next] != c) next++
    for (let i = pre + 1; i <= next; i++) {
      ans[i] = Math.min(i - pre, next - i)
    }
    pre = next
    next++
  }
  for (let i = s.length - 1; i >= last; i--) {
    ans[i] = i - last
  }
  return ans
};

File: 821-shortest-distance-to-a-character/821-shortest-distance-to-a-character.go
func dist(pre, i, next int) int {
  dp, dn := i - pre, next - i
  if dp > dn {
    return dn
  } else {
    return dp
  }
}

func shortestToChar(s string, c byte) []int {
  ans := make([]int, len(s))
  pre := 0
  next := -len(s)
  j := 0
  for i := 0; i < len(s); i++ {
    if s[i] == c {
      pre = next
      if pre >= 0 {
        ans[i] = 0
        j = pre+1
      }
      next = i
      for ;j < next; j++ {
        ans[j] = dist(pre, j, next)        
      }
    }
  }
  ans[next] = 0
  for j = next+1; j < len(s); j++ {
    ans[j] = j - next
  }
  return ans
}